### PR TITLE
DATALAB-291: Connect Anamita to Elasticsearch

### PR DESCRIPTION
**Pull request by twindebank**
Added Elasticsearch client to API. Configure using environmental variables (as per readme)
Have significantly refactored the code:
- any client-specific code lies in the `clients` directory
- tests refactored to use large samples of real data (cover more edge cases)
- SPARQL client not covered by tests yet (fully), old tests need refactoring to fit with new structure. I chose to create PR anyway as this isn't priority and is a good few hours more of work.

### CODE REVIEW

File: app/apiparams/lists.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+"""
+Lists of API Parameters for particular endpoints.
+It accepts the parameter definitions from the clients as each client may have a different implementation of the same
+parameter.
+"""
```
Comment (jontutcher) [inline app/apiparams/lists.py]: This could probably be a bit clearer

**Comment by jontutcher** [comment]
|`app/apiparams/lists.py:21`|
|---|
Minor point, but probably more pythonic to replace these `elif`s with `if`s, seeing as you're returning from the cases above anyway :)

File: app/apiparams/types.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+"""Definitions of types that may be used to validate and format incoming query parameters."""
+import iso8601
+from rdflib import URIRef, Literal
+
+from app.clients.sparql.namespaces import namespaces as NS
+
+
+class BoolFromString:
```
Comment (jontutcher) [inline app/apiparams/types.py]: Is this class used anywhere

File: app/apiparams/types.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    def __new__(cls, bool_str):
+        if bool_str not in ['true', 'True', 'false', 'False']:
+            raise ValueError(f'"{bool_str}" is not a valid boolean string')
+        real_bool = True if bool_str in ['true', 'True'] else False
+        return real_bool
+
+
+class MediaUriRef(URIRef):
```
Comment (jontutcher) [inline app/apiparams/types.py]: |`line 16`|
|---|
These classes would benefit from docstrings, as it's not obvious what they're for at first glance :)

**Comment by jontutcher** [comment]
In general, it might be good to refactor your unit tests to be a bit more understandable. I think @tatiana might have some good tips here, but it'd be good if the names of the test files corresponded with the modules they test (so you can immediately find tests for a particular method). In the cases where you've tested the same method multiple times, having all those tests in an appropriately-named module might work!

**Comment by jontutcher** [comment]
|`app/clients/client_interface.py:124`|
|---|
could this be named more clearly?

**Comment by jontutcher** [comment]
|`app/clients/client_interface.py:129`|
|---|
Seeing in this context I understand what the method does now. I think it might be worth renaming this to `validate_parameter_and_cast_to_rdflib`, or maybe even call two separate methods: `.cast_to_rdflib` and `.validate`

**Comment by jontutcher** [comment]
|`app/clients/elastic/process_response.py:4`|
|---|
`map_search_results_to_api_view` or something? naming things is hard!

**Comment by jontutcher** [comment]
|`app/clients/elastic/process_response.py:6`|
|---|
is `clip` == one entry in the elasticsearch resultset?

**Comment by jontutcher** [comment]
|`app/clients/elastic/process_response.py:9`|
|---|
You could factor this innards of this loop into `map_search_result_to_python` or something to make `transform_hits` shorter - not necessary at all but a suggestion :)

File: app/clients/elastic/process_response.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+from app.utils.conversions import lower_camel_case_to_upper
+
+
+def transform_hits(es_res):
+    """Transform raw ES results to align with API specification."""
+    hits = [clip for clip in es_res['hits']['hits']]
+    clips = []
```
Comment (jontutcher) [inline app/clients/elastic/process_response.py]: It might be more readable to use `filter()` here instead?

File: app/api.py [hunk 5]
```diff
@@ -1,28 +1,50 @@
...
+    'elasticsearch': ESClient
+}
 
-DEFAULT_HTTP_PORT = "5001"
-PORT = int(os.getenv("PORT", DEFAULT_HTTP_PORT))
+PORT = int(os.getenv("PORT", constants.DEFAULT_HTTP_PORT))
+DB_ENDPOINT = os.getenv('DB_ENDPOINT', constants.DEFAULT_DB_ENDPOINT)
+DB_USER = os.getenv('DB_USER', constants.DEFAULT_DB_USER)
+DB_PASS = os.getenv('DB_PASS', constants.DEFAULT_DB_PASS)
+DB_CLIENT = db_clients[os.getenv('DB_CLIENT', constants.DEFAULT_DB_CLIENT)]
```
Comment (tatiana) [inline app/api.py]: The client instance is not a constant and I'd suggest it is called just `db_client_class` and not `DB_CLIENT`. It is important to state this is the class - otherwise it seems it is the client instance which could be used straightaway.

File: app/apiparams/lists.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+"""This file defines which parameters are used/needed for each endpoint. DB Clients to be imoplemented should
+implement a `parameter_definitions` property that contains definitions of all the parameters listed here.
+
+Returns objects to map parameter values to the correct format, given all of the mapping objects defined by the client,
+and the relevant endpoint. """
+
+
+def get_param_mappers_for_endpoint(endpoint, definitions):
+    if endpoint == 'content':
```
Comment (tatiana) [inline app/apiparams/lists.py]: It helps debugging and can improve readability if we asign the dictionaries to a variable and return it after - this way we don't get lost with several lines related to the return statement. This is style, and I'm fine with what you decide :)

File: app/apiparams/mapping.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
+from app.apiparams.lists import get_param_mappers_for_endpoint
+from exceptions.queryexceptions import InvalidInputParameterValue, InvalidInputParameter, InvalidInputQuery
+
+
+class ParameterMapper:
```
Comment (tatiana) [inline app/apiparams/mapping.py]: It feels the classes themselves should handle their own mapping and be self contained. It is unclear the benefit of having an external class which will handle other classes mapping.

File: app/apiparams/mapping.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+                v_l = [v] if not isinstance(v, list) else v
+                v_cast = [self.param_type(i, **self.kwargs) for i in v_l]
+            else:
+                v_cast = self.param_type(v, **self.kwargs)
+        except Exception as e:
+            raise InvalidInputParameterValue(f'Value(s) {v} incorrectly formatted: {str(e)}.')
+        return v_cast
+
+
+def map_param_values_to_db_compatible(query_params, endpoint, parameter_definitions):
```
Comment (tatiana) [inline app/apiparams/mapping.py]: Again, @twindebank, I strongly suggest our client classes are self contained and handle this type of thing as methods.

File: app/clients/client_superclass.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    def get_similar(self, validated_item_uri, validated_query_params):
+        """
+        Get list of content related to given item uri and filter based on `query_params`.
+        Args:
+            validated_query_params: _validated_ query parameters for filtering results
+            validated_item_uri (rdflib.URIRef): _validated_ uri of item to get data for
+        """
+
+    # parameter mapping
+    def map_content_query_params_to_db_compatible(self, query_params):
```
Comment (tatiana) [inline app/clients/client_superclass.py]: I strongly suggest these methods are not implemented in this class, but in the children classes:
- map_content_query_params_to_db_compatible
- map_item_query_uri_to_db_compatible
- map_similar_query_params_to_db_compatible

It should not be part of the responsaiblity of the parent class to handle specific cases of each implementation (eg. define how a parameter maps to a query in a specific db).

File: app/clients/elastic/querybuilder/get_content.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        raise NotImplementedError('The parameter `publishedAfter` is not yet implemented')
+    if region is not None:
+        raise NotImplementedError('The parameter `region` is not yet implemented')
+    if similarity_method is not None:
+        raise NotImplementedError('The parameter `similarityMethod` is not yet implemented')
+    if sort is not None and random:
+        raise InvalidInputParameterCombination('Cannot specify both `sort` and `random`.')
+
+    search = Search(index='pips')
+    search = search[offset:offset + limit]  # TODO: THIS DOESNT WORK?? query builds as should but no effect
```
Comment (tatiana) [inline app/clients/elastic/querybuilder/get_content.py]: ElasticSearch doesn't know abou offsets and limits - it calls the properties from and size.

It seems the syntax would be: `Search(index='pips').extra(from_=offset, size=limit)`

https://github.com/elastic/elasticsearch-dsl-py/blob/e75ac158e26bc1981ca89e8033911cb21940e3d9/elasticsearch_dsl/search.py

File: app/clients/elastic/querybuilder/get_similar.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    URI."""
+    if published_after is not None:
+        raise NotImplementedError('The parameter `publishedAfter` is not yet implemented.')
+    if region is not None:
+        raise NotImplementedError('The parameter `region` is not yet implemented.')
+    if similarity_method is not None:
+        raise NotImplementedError('The parameter `similarityMethod` is not yet implemented for ES.')
+
+    search = Search(index='pips')
+    search = search[offset:offset + limit]  # TODO: THIS DOESNT WORK?? query builds as should but no effect
```
Comment (tatiana) [inline app/clients/elastic/querybuilder/get_similar.py]: ElasticSearch doesn't know abou offsets and limits - it calls the properties from and size.

It seems the syntax would be: `Search(index='pips').extra(from_=offset, size=limit)`

https://github.com/elastic/elasticsearch-dsl-py/blob/e75ac158e26bc1981ca89e8033911cb21940e3d9/elasticsearch_dsl/search.py

File: app/utils/constants.py [hunk 12]
```diff
@@ -3,9 +3,12 @@
 DEFAULT_QUERY_LIMIT = 20
 DEFAULT_QUERY_OFFSET = 0
 
-DEFAULT_STARDOG_ENDPOINT = 'http://localhost:5820/content-graph/query'
-DEFAULT_STARDOG_USER = 'admin'
-DEFAULT_STARDOG_PASS = 'admin'
+DEFAULT_DB_ENDPOINT = 'http://localhost:5820/content-graph/query'
```
Comment (tatiana) [inline app/utils/constants.py]: Given our architectural decision, I'd suggest we set the default to ElasticSearch.

File: tests/test_apiparams_types.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+from iso8601 import iso8601
+
+from app.apiparams import types
+from app.apiparams.types import BoolFromString, MediaUriRef, LowercaseLiteral, ValidatedDatetime, StrictlyPositiveInt, \
+    URIStr, LowercaseStr
+
+
+# test parameter types
+def test_bool_from_string():
+    bool = BoolFromString('true')
```
Comment (tatiana) [inline tests/test_apiparams_types.py]: Perhaps we should avoid redefining the `bool` type, even in the context of a test ;)
In this case we could just use the asserts straight away, eg.
`assert BoolFromString('true')`

File: tests/test_clients_elastic/test_clients_elastic_process_response.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+import json
+
+import app.api as api
+from app.clients.elastic.client import ESClient
+
+
+def test_content_endpoint_map_hits_to_api_response(flask_app, monkeypatch):
+    # dont test parameters here just test content processing
+    monkeypatch.setattr(api, 'DB_CLIENT', ESClient)  # instead of using env vars to select db
+    monkeypatch.setattr(ESClient, 'setup_connection', lambda *_, **__: None)
```
Comment (tatiana) [inline tests/test_clients_elastic/test_clients_elastic_process_response.py]: I suggest we use `unittest.mock` to improve the isolation and readability. We've been successfully using it with Pytest in other projects. The `patch` decorator can help on this.

File: tests/test_clients_elastic/test_clients_elastic_querybuilder_get_content.py [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+        'from': 10,
+        'size': 10
+    }
+
+    #test random param
+    val_params = {k: v['validated'] for k, v in params.items() if
+                  k not in non_implemented_params and k in expected_params and k != 'sort'}
+    monkeypatch.setattr(get_content, 'randint', lambda *_: 1)
+    body = build_query_body(**val_params)
+    assert body == {
```
Comment (tatiana) [inline tests/test_clients_elastic/test_clients_elastic_querybuilder_get_content.py]: it can improve readability to have the dictionary assigned to a variable before (eg. `expected`) and have a very short assert line, eg:
`assert body == expected`

File: tests/test_clients_elastic/test_clients_elastic_querybuilder_get_similar.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+
+# response processing
+from exceptions.queryexceptions import InvalidInputParameterCombination
+
+item_uri = 'programmes:bbc.co.uk,2018/FIXME/p05q11tt'
+
+# query building
+def test_query_building_no_params():
+    body = build_query_body(item_uri=item_uri)
+    assert body == {'query': {'bool': {'should': [{'more_like_this': {
```
Comment (tatiana) [inline tests/test_clients_elastic/test_clients_elastic_querybuilder_get_similar.py]: it can improve readability to have the dictionary assigned to a variable before (eg. `expected`) and have a very short assert line, eg:
`assert body == expected`

File: tests/test_clients_elastic/test_clients_elastic_querybuilder_get_similar.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+
+def test_query_building_all_implemented_params():
+    params = json.load(open("test_clients_elastic/data/param_examples.json"))
+    expected_params = list(inspect.signature(build_query_body).parameters)
+
+    val_params = {k: v['validated'] for k, v in params.items() if
+                  k not in non_implemented_params and k in expected_params}
+    body = build_query_body(**val_params)
+    assert body == {'query': {'bool': {'filter': [{'term': {'mediaType': 'audio'}}, {'term': {'mediaType': 'video'}}, {'range': {'duration': {'lte': 100}}}], 'should': [{'more_like_this': {'like': {'_index': 'pips', '_type': 'clip', '_id': 'http://exampleuri.com'}, 'fields': ['title', 'masterBrand.mid', 'mediaType'], 'min_term_freq': 1, 'min_doc_freq': 1}}, {'nested': {'path': 'genres', 'query': {'more_like_this': {'fields': ['genres.key'], 'like': {'_index': 'pips', '_type': 'clip', '_id': 'http://exampleuri.com'}, 'min_term_freq': 1, 'min_doc_freq': 1}}}}], 'minimum_should_match': 1}}, 'from': 10, 'size': 10}
```
Comment (tatiana) [inline tests/test_clients_elastic/test_clients_elastic_querybuilder_get_similar.py]: it can improve readability to have the dictionary assigned to a variable before (eg. `expected`) and have a very short assert line, eg:
`assert body == expected`

File: tests/test_clients_elastic/test_clients_elastic_query_parameter_mappers.py [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+def test_parameter_mappers(param):
+    """Test each parameter mapper with good and bad inputs"""
+    mappers = query_parameter_mappers
+    param_name, param_vals = param
+    validator = getattr(mappers, param_name)
+
+    validated_param = validator.validate(param_vals['good_raw'])
+    assert validated_param == param_vals['validated']
+
+    if param_vals['bad_format_raw'] is not None:
```
Comment (tatiana) [inline tests/test_clients_elastic/test_clients_elastic_query_parameter_mappers.py]: Tests should be as simple as possible - otherwise we'd need to write tests for tests.
Whenever I see an `if` or `for`, it raises a red flag. Please, refactor this.

File: tests/test_clients_elastic/test_clients_elastic_querybuilder_get_content.py [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+    with pytest.raises(InvalidInputParameterCombination):
+        build_query_body(sort='', random=True)
+
+
+def test_build_query_body_all_implemented_params(monkeypatch):
+    params = json.load(open("test_clients_elastic/data/param_examples.json"))
+    expected_params = list(inspect.signature(build_query_body).parameters)
+
+    # test with sort param separately to random param (cant specify both)
+    val_params = {k: v['validated'] for k, v in params.items() if
```
Comment (tatiana) [inline tests/test_clients_elastic/test_clients_elastic_querybuilder_get_content.py]: Tests should be as simple as possible - otherwise we'd need to write tests for tests.
Whenever I see an if or for, it raises a red flag. Please, refactor this.

To me, this would be a blocker to approve the PR. Let me know if you want to talk about strategies to refactor it.

File: tests/test_clients_elastic/test_clients_elastic_querybuilder_get_content.py [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+        'sort': ['duration', 'publicationDate', 'masterBrand',
+                 {'duration': {'order': 'desc'}},
+                 {'publicationDate': {'order': 'desc'}},
+                 {'masterBrand': {'order': 'desc'}}],
+        'from': 10,
+        'size': 10
+    }
+
+    #test random param
+    val_params = {k: v['validated'] for k, v in params.items() if
```
Comment (tatiana) [inline tests/test_clients_elastic/test_clients_elastic_querybuilder_get_content.py]: Same here. We should not have this in the test.

File: README.md [hunk 21]
```diff
@@ -12,21 +12,30 @@ python3 -m venv env
...
 ```
 
-### 2. Run the service:
+### 2b. Run with Elastic
+```
+export DB_CLIENT='elasticsearch'
+export DB_ENDPOINT=http://localhost:5001
+export DB_USER=$USERNAME
+export DB_PASS=$PASSWORD
+```
```
[the remark this answers is not in the document]
Comment (twindebank) [reply README.md]: They are mutually excusive yes (for now) - I'll make the readme clearer as I've unintentonally made it look like they should be run together.

File: app/api.py [hunk 22]
```diff
@@ -1,28 +1,50 @@
...
+logger.info(f'Using credentials:\n endpoint: {DB_ENDPOINT}\n user: {DB_USER}\n pass: {DB_PASS}')
 
 app = Flask(__name__)
 
 
+def get_client():  # pragma: no cover
+    # this is covered by tests but coverage.py doesnt pick it up
```
[the remark this answers is not in the document]
Comment (twindebank) [reply app/api.py]: I can't see how the `test_get_client_and_close` test in `test_api.py` doesn't cover it? `g` and the `DB_CLIENT` object are patched, but I think it should still show as covered? I'll do the refactoring mentioned below and see if the problem persists.

File: README.md [hunk 23]
```diff
@@ -12,21 +12,31 @@ python3 -m venv env
...
+export DB_ENDPOINT=http://$SERVER:$PORT/content-graph-test/query
+export DB_USER=$USERNAME
+export DB_PASS=$PASSWORD
 ```
 
-### 2. Run the service:
+#### For use with Elastic
+```
+export DB_CLIENT='elasticsearch'
+export DB_ENDPOINT=http://localhost:5001
```
Comment (tatiana) [inline README.md]: Perhpas we should set the default port to 9200, the default ES port

File: app/clients/client_superclass.py (hunk 9 above)
Comment (twindebank) [reply app/clients/client_superclass.py]: These would be the same for each child class and I don't want to repeat it for each child, so instead I have factored it out into `mapping.py`

File: app/clients/elastic/querybuilder/get_similar.py (hunk 11 above)
Comment (twindebank) [reply app/clients/elastic/querybuilder/get_similar.py]: This is the way to implement offset and limits using `elasticsearch-dsl`. For example, 
```
search = Search(index='pips')
search = search[10:25]
```
creates a search body which looks like:
```
{'query': {'match_all': {}}, 'from': 10, 'size': 15}
```
I believe that is the correct syntax, yet it still doesnt work?

**Comment by twindebank** [comment]
@tatiana @jontutcher - I have responded to the changes, perhaps it needs another quick review then its good to go? Thanks for the excellent reviewing so far, appreciate the effort that goes in 👌

File: app/api.py [hunk 24]
```diff
@@ -1,28 +1,54 @@
...
 
 from SPARQLWrapper.SPARQLExceptions import QueryBadFormed
-from flask import Flask, jsonify, request
-import os
-import logging
-
-from app import contentgraph
-from exceptions.clientexceptions import DBClientResponseError, NoResultsFoundError
-from exceptions.helpers import log_last_exception, format_traceback_as_html
+from flask import Flask, jsonify, request, g
```
Comment (tatiana) [inline app/api.py]: suggestion: alphabetic order (`flask` before `SPARQLWrapper`)

File: app/clients/elastic/process_response.py [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+
+def map_hits_to_api_spec(es_res):
+    """Transform raw ES results to align with API specification."""
+    hits = es_res['hits']['hits']
+    clips = []
+    for hit in hits:
+        clip = copy(hit['_source'])
+        clip['uri'] = hit['_id']
+        clip['masterBrand'] = clip['masterBrand']['mid'] if clip['masterBrand'] is not None else None
+        genres = {'topLevel': [], 'secondLevel': [], 'thirdLevel': []}
```
Comment (tatiana) [inline app/clients/elastic/process_response.py]: Hey @twindebank I've just realised we're doing this conversion. I don't think we need this - we had aligned on the mapping in ES with @janae and I believe we're fine to represent the data in the way we structured at ES - I'd suggest we simply update the API spec. I believe we should avoid as much as possible this kind of internal transformation - and have the document at ES mapping how we want to represent our data as much as possible.

File: app/api.py (hunk 24 above)
Comment (twindebank) [reply app/api.py]: Interesting, PyCharm `Optimise Imports` wants it the other way around. Looks like PEP8 doesn't seem to specify order. I'm happy to change it here but I'm not sure I'll remember to keep it up through the rest of my code as I usually rely on PyCharm to quickly rejigger it all for me

File: app/clients/elastic/process_response.py (hunk 25 above)
Comment (twindebank) [reply app/clients/elastic/process_response.py]: also, if we changed the api spec that would mean having to change the ES schema and reingesting all the data

Comment (tatiana) [reply app/clients/elastic/process_response.py]: Let's have this conversation in a separate moment, I'm fine leaving things as they are for now. It feels it would be possible for us to have in ES a model which is very close to what we need, without needing to convert this model in the content graph. There will be situations when conversions will be needed - it just doesn't feel the case here. I undestand we need to align ingestion etc - so let's discuss this separately.

File: app/api.py (hunk 24 above)
Comment (tatiana) [reply app/api.py]: Yes, ordering is not in PEP8, but I really feel it is worth to organise things alphabetically. This helps readability.

File: app/clients/elastic/process_response.py (hunk 25 above)
Comment (twindebank) [reply app/clients/elastic/process_response.py]: okay sure :)

**Comment by tatiana** [comment]
Hey, @twindebank, let's address the pagination issue as soon as the deployment issue is solved. In order to follow the deployment pipeline, I'm fine with merging this to master.

**Comment by jontutcher** [comment]
This looks good! Happy to accept on the condition that we spend an hour or so doing a code share / refactor session for readability, because I think it's really important for this code to be as understandable as possible as it gets more complex :) If you're up for this @twindebank @tatiana I'll put a 1 point ticket in the next sprint!
